index.rst

```diff
@@ -1,61 +1,311 @@
+
+Simplification of wfs process commands
+--------------------------------------
+
+As mention above, the wfs process commands receives file names and path.
+With the data already ingested in the butler the images can be uniquely identified by the ``visitId`` entry in the database.
+That means, all process commands need to be updated to receive the ``visitId`` instead of the files name and path.
+
+At the same time, most of the metadata that is now passed as part of the process commands (``fieldRA``, ``fieldDEC``, ``filter``, ``cameraRotation``) is also available as metadata in the butler and can be removed from the command payload.
+
+Furthermore, through the ``visitId`` one can identify the source of the image (ComCam or LSSTCam), so there is no need for special commands for each one of them.
+
+At the same time, we can differentiate between an Intra/Extra wavefront analysis process and a corner wavefront sensors by the input alone.
+If only one input ``visitId`` then it can be assumed to be a corner wavefront sensor run.
+If, in addition to ``visitId`` an ``extraId`` is given, it can be considered and Intra/Extra wavefront process.
+
+The command would be used as follows:
+
+.. code-block:: py
+
+  from lsst.ts import salobj
+
+  d = salobj.Domain()
+
+  mtaos = salobj.Remote(d, "MTAOS")
+
+  await mtaos.start_task
+
+  # Process full array intra/extra focal images.
+  # Frames can be ComCam or LSSTCam
+  await mtaos.cmd_runWEP.set_start(
+      visitId=2020103000040, extraId=2020103000041
+  )
+
+  # Process corner wavefront sensor
+  # Frames should be LSSTCam only. If ComCam command will raise an exception.
+  await mtaos.cmd_runWEP.set_start(
+      visitId=2020103000040
+  )
+
+Source selection command
+------------------------
+
+In most cases, we will known the coordinates of the next visit field ahead of time.
+That means some pre-processing activities can be done in parallel, while the telescope slew and the image is acquired.
+One of these processes is to prepare a list of suitable targets for wavefront sensing.
+
+This could be done by adding a command that would cause the MTAOS to run the source selection routine.
+The command payload would minimally be;
+
+  - RA: Right ascension of the field (in hours).
+  - Dec: Declination of the field (in degrees).
+  - PA: Position angle (in degrees).
+  - mode: An enumeration specifying the wfs mode; corner, full array, etc.
+
+Issuing the command would look something like:
+
+.. code-block:: py
+
+  from lsst.ts.idl.MTAOS import WFSMode
+
+  # Select sources for Main Camera Corner wavefront sensor mode
+  await mtaos.cmd_selectSources.set_start(ra=14., dec=-10., pa=0., type=WFSMode.Corner)
+
+  # Select sources for Main Camera Full array mode
+  await mtaos.cmd_selectSources.set_start(ra=14., dec=-10., pa=0., type=WFSMode.FullArray)
+
+  # Select sources for ComCam mode
+  await mtaos.cmd_selectSources.set_start(ra=14., dec=-10., pa=0., type=WFSMode.ComCam)
```

Comment:
I think the argument of type is not needed. The `MTAOS` read the configuration file in `ts_config_mttcs` now:
https://github.com/lsst-ts/ts_config_mttcs/blob/develop/MTAOS/v1/default.yaml#L7

Reply:
The `type` tells us the geometry of the CCDs. The source selector's job is to determine what stars fall onto these CCD, right?

Reply:
Yes. The configuration file in `ts_config_mttcs` has assigned the `camera` type to use. Therefore, I do not prefer the user could change the `type` in the run time (this means the `type` here = `camera` in the configuration file). The parameter of `camera` in the configuration file will read a lot of related parameters.

Reply:
Having the enumeration would allow us to run the MTAOS with the different modes without need to reconfigure the CSC. This is specially useful if you want to run full array or corner wavefront sensor.

Reply:
The problem is that the `camera` or `type` will need the `ts_wep` to re-load a lot of related parameters. This means the `wep` instance hold by `Model` instance will need to redo the instantiation for each time of change of `camera` or `type`. In addition, the `ts_ofc` has the similar problem as well. This is why you can see the `instrument` in the configuration file as well:

https://github.com/lsst-ts/ts_config_mttcs/blob/develop/MTAOS/v1/default.yaml#L11

Therefore, if you really want to be able to change the `camera` and `type` in the run time, you may need to add another argument of `instrument` for the `ts_ofc` to use. In addition, at each time of change, the `MTAOS` will be very slow because the instances of `wep` and `ofc` hold by `Model` will need to do the re-instantiation (>10 sec).